### PR TITLE
Fix #1080: Protection against duplicate entries in the pa_operation_template table

### LINKED ISSUE #1080: Protection against duplicate entries in the pa_operation_template table
It is now possible to create duplicate records in the pa_operation_template table. It could cause some errors.

On API level there were already a validation implemented, however there wasn't an unique constraint on database level. It should not be possible to create a template on API level (`/rest/v3/operation/template/create`) nor on database level (table `pa_operation_template`) when the name is not unique.

### PR DESCRIPTION
**Pull request by jnpsk**
Fix #1080
- OperationTemplateEntity.templateName set to unique
- Created liquibase changeset
- Tests

### CODE REVIEW

File: docs/db/changelog/changesets/powerauth-java-server/1.5.x/20231018-add-constraint-operation-template-name.xml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-4.9.xsd">
+
+    <changeSet id="1"  logicalFilePath="powerauth-java-server/1.5.x/20231018-add-constraint-operation-template-name.xml" author="Jan Pesek">
```
Comment (banterCZ) [inline docs/db/changelog/changesets/powerauth-java-server/1.5.x/20231018-add-constraint-operation-template-name.xml:6]:
```suggestion
    <changeSet id="1" logicalFilePath="powerauth-java-server/1.5.x/20231018-add-constraint-operation-template-name.xml" author="Jan Pesek">
```
Moreover, the version 1.5.0 is already released and should be targeted into 1.6.x

File: docs/db/changelog/changesets/powerauth-java-server/1.5.x/20231018-add-constraint-operation-template-name.xml [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+        <preConditions onFail="MARK_RAN">
+            <tableExists tableName="pa_operation_template" />
+            <columnExists tableName="pa_operation_template" columnName="template_name" />
+        </preConditions>
+        <comment>Add unique constraint to pa_operation_template.template_name</comment>
+        <addUniqueConstraint tableName="pa_operation_template" columnNames="template_name" />
```
Comment (banterCZ) [inline docs/db/changelog/changesets/powerauth-java-server/1.5.x/20231018-add-constraint-operation-template-name.xml:12]: Should follow our database naming standards https://github.com/wultra/development-internal/wiki/Database-Naming#unique-constraints

File: docs/db/changelog/changesets/powerauth-java-server/1.5.x/20231018-add-constraint-operation-template-name.xml [hunk 3]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-4.9.xsd">
+
+    <changeSet id="1"  logicalFilePath="powerauth-java-server/1.5.x/20231018-add-constraint-operation-template-name.xml" author="Jan Pesek">
+        <preConditions onFail="MARK_RAN">
```
Comment (banterCZ) [inline docs/db/changelog/changesets/powerauth-java-server/1.5.x/20231018-add-constraint-operation-template-name.xml:7]: This check should be based on the presence of the constraint (not to fail if anybody has applied it manually before, for example).

File: powerauth-java-server/src/test/java/io/getlime/security/powerauth/app/server/database/repository/OperationTemplateRepositoryTest.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    private OperationTemplateRepository repository;
+
+    @Autowired
+    private TestEntityManager entityManager;
+
+    @Test
+    void testDuplicateOperationTemplateCreation() {
+        final String templateName = "login";
+
+        repository.save(createOperationTemplateEntity(templateName));
```
Comment (banterCZ) [inline powerauth-java-server/src/test/java/io/getlime/security/powerauth/app/server/database/repository/OperationTemplateRepositoryTest.java:53]: FYI You may prepare data using `@Sql`.

File: powerauth-java-server/src/test/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/OperationTemplateServiceBehaviorTest.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+class OperationTemplateServiceBehaviorTest {
+
+    @Autowired
+    private OperationTemplateServiceBehavior service;
+
+    @Test
+    void testDuplicateOperationTemplateCreation() throws Exception {
+        final String templateName = "login";
+
+        service.createOperationTemplate(createOperationTemplateCreateRequest("login"));
```
Comment (banterCZ) [inline powerauth-java-server/src/test/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/OperationTemplateServiceBehaviorTest.java:45]: Didn't you want to use `templateName`? Anyway, what is the reason to test both repository and service?
```suggestion
        service.createOperationTemplate(createOperationTemplateCreateRequest(templateName));
```

**Review by banterCZ — COMMENTED** [review]
Let's move the liquibase to the next version, please. Otherwise good job.

File: powerauth-java-server/src/test/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/OperationTemplateServiceBehaviorTest.java (hunk 5 above)
Comment (jnpsk) [reply powerauth-java-server/src/test/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/OperationTemplateServiceBehaviorTest.java:45]: > Anyway, what is the reason to test both repository and service?

The `OperationTemplateServiceBehavior.createOperationTemplate()` checks uniqueness on service level by querying a template by its name first. If a template with the same name already exists, it throws an exception without triggering database constraint. Although it means the template creation using the API is not influenced by this PR, I wanted to cover it with a test, as it is closely related.

This issue address direct database insertion, which must be handled by the database constraint and is therefore tested on the repository level.

File: powerauth-java-server/src/test/java/io/getlime/security/powerauth/app/server/database/repository/OperationTemplateRepositoryTest.java (hunk 4 above)
Comment (jnpsk) [reply powerauth-java-server/src/test/java/io/getlime/security/powerauth/app/server/database/repository/OperationTemplateRepositoryTest.java:53]: Thanks for the hint! However this is a simple one-line insertion, so I prefer this solution over including a sql file.

File: docs/db/changelog/changesets/powerauth-java-server/1.6.x/20231018-add-constraint-operation-template-name.xml [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-4.9.xsd">
+
+    <changeSet id="1" logicalFilePath="powerauth-java-server/1.6.x/20231018-add-constraint-operation-template-name.xml" author="Jan Pesek">
+        <preConditions onFail="MARK_RAN">
+            <not>
+                <uniqueConstraintExists tableName="pa_operation_template" columnNames="template_name" constraintName="pa_operation_template_template_name_uk" />
+            </not>
+        </preConditions>
+        <comment>Add unique constraint to pa_operation_template.template_name</comment>
+        <addUniqueConstraint tableName="pa_operation_template" columnNames="template_name" constraintName="pa_operation_template_template_name_uk" />
```
Comment (banterCZ) [inline docs/db/changelog/changesets/powerauth-java-server/1.6.x/20231018-add-constraint-operation-template-name.xml:13]: @zcgandcomp Should we keep recording the db changes also in the migration guides or not?

Comment (romanstrobl) [reply docs/db/changelog/changesets/powerauth-java-server/1.6.x/20231018-add-constraint-operation-template-name.xml:13]: We should add information into the migration guide, please add this information in the usual way as seen in previous migration guides.

**Review by romanstrobl — APPROVED** [review]
NIce improvement, however we should also add step to the migration guide. We should also document how to proceed in case addition of the unique constraint fails.

File: docs/db/changelog/changesets/powerauth-java-server/1.6.x/20231018-add-constraint-operation-template-name.xml (hunk 6 above)
Comment (banterCZ) [reply docs/db/changelog/changesets/powerauth-java-server/1.6.x/20231018-add-constraint-operation-template-name.xml:13]: Should we be concrete about the changes or just mention that some changes are needed and link a script, which will be generated during the release?

Comment (zcgandcomp) [reply docs/db/changelog/changesets/powerauth-java-server/1.6.x/20231018-add-constraint-operation-template-name.xml:13]: https://github.com/wultra/development-internal/wiki/Using-Liquibase#13-migration-between-versions. Yes please add a migration guide and mention the change. In theory it is possible to have (invalid) duplicities in the DB. The migration script in SQL will be generated during release procedure.

**Review by zcgandcomp — CHANGES_REQUESTED** [review]

**Review by banterCZ — APPROVED** [review]
The migration guide is good for me.

**Pull request merged**